File: F#/BowlingGameTests/GameTests.fs
module GameTests

open System
open Xunit
open Game

let assert_rolls_makes_score rolls score =
    let game = rolls |> List.fold Game.roll []
    Assert.Equal(score, Game.score game)

[<Fact>]
let ``All gutters scores 0`` () =
    assert_rolls_makes_score [for i in 1..20 -> 0] 0

[<Fact>]
let ``All ones scores 20`` () =
    assert_rolls_makes_score [for i in 1..20 -> 1] 20

[<Fact>]
let ``Spare counts next roll twice`` () =
    assert_rolls_makes_score [4;6;3] 16

[<Fact>]
let ``Strike counts next two rolls twice`` () =
    assert_rolls_makes_score [10;4;3] 24

[<Fact>]
let ``Perfect game scores 300`` () =
    assert_rolls_makes_score [for i in 1..12 -> 10] 300

File: C#/BowlingGameTests/GameTests.cs
using BowlingGame;
using Xunit;

namespace BowlingGameTests
{
    public class GameTests
    {
        private Game _game;

        public GameTests()
        {
            _game = new Game();
        }

        private void RollMany(int pins, int rolls)
        {
            for (int i = 0; i < rolls; i++)
            {
                _game.Roll(pins);
            }
        }

        [Fact]
        public void All_Gutters_Scores_0()
        {
            // Act

            // Assert
            RollMany(0, 20);

            // Assert
            Assert.Equal(0, _game.Score());
        }

        [Fact]
        public void All_Ones_Scores_20()
        {
            // Arrange

            // Act
            RollMany(1, 20);

            // Assert
            Assert.Equal(20, _game.Score());
        }

        [Fact]
        public void Spare_Doubles_Next_Roll()
        {
            // Arrange
            _game.Roll(4);
            _game.Roll(6);

            // Act
            _game.Roll(3);

            // Assert
            Assert.Equal(16, _game.Score());
        }
    }
}